contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,620 @@
+        "favorite_number": 7,
+        "thanks_to": "thockin"
+      }
+    }
+  ]
+}
+```
+
+### Enable/Disable
+
+When specified, the list of plugins for a particular extension point are the
+only ones enabled. If an extension point is omitted from the config, then the
+default set of plugins is used for that extension point.
+
+### Change Evaluation Order
+
+When relevant, plugin evaluation order is specified by the order the plugins
+appear in the configuration. A plugin that registers for multiple extension
+points can have different ordering at each extension point.
+
+### Optional Args
+
+Plugins may receive arguments from their config with arbitrary structure.
+Because one plugin may appear in multiple extension points, the config is in a
+separate list of `PluginConfig`.
+
+For example,
+
+```json
+{
+   "name": "ServiceAffinity",
+   "args": {
+      "LabelName": "app",
+      "LabelValue": "mysql"
+   }
+}
+```
+
+```go
+func NewServiceAffinity(args json.RawMessage, h PluginHandle) (Plugin, error) {
+    var config struct {
+        LabelName, LabelValue string
+    }
+    if err := json.Unmarshal(args, &config); err != nil {
+        return nil, errors.Wrap(err, "could not parse args")
+    }
+    //...
+}
+```
+
+### Backward compatibility
+
+The current `KubeSchedulerConfiguration` kind has `apiVersion:
+kubescheduler.config.k8s.io/v1alpha1`. This new config format will be either
+`v1alpha2` or `v1beta1`. When a newer version of the scheduler parses a
+`v1alpha1`, the "policy" section will be used to construct an equivalent plugin
+configuration.
+
+*Note: Moving `KubeSchedulerConfiguration` to `v1` is outside the scope of this
+design, but see also
+https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md
+and https://github.com/kubernetes/community/pull/3008*
+
+# USE CASES
+
+These are just a few examples of how the scheduling framework can be used.
+
+## Coscheduling
+
+Functionality similar to
+[kube-batch](https://github.com/kubernetes-sigs/kube-batch) (sometimes called
+"gang scheduling") could be implemented as a plugin. For pods in a batch, the
+plugin would "accumulate" pods in the [permit](#permit) phase by using the
+"wait" option. Because the permit stage happens after [reserve](#reserve),
+subsequent pods will be scheduled as if the waiting pod is using those
+resources. Once enough pods from the batch are waiting, they can all be
+approved.
+
+## Dynamic Resource Binding
+
+[Topology-Aware Volume Provisioning](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/)
+can be (re)implemented as a plugin that registers for [filter](#filter) and
+[pre-bind](#pre-bind) extension points. At the filtering phase, the plugin can
+ensure that the pod will be scheduled in a zone which is capable of provisioning
+the desired volume. Then at the pre-bind phase, the plugin can provision the
+volume before letting scheduler bind the pod.
```

Comment:
this is interesting, would this replace the topology aware provisioning ?

Reply:
Yes, one of our goals with the framework is to convert that feature into a plugin. You've probably encountered the volume binder logic in a number of surprising places. Pulling that all into a plugin will seriously clean up our business logic.

Also, the new pre-bind extension point would allow volume binder to handle pods with one scheduling cycle instead of the two that it takes today.

Reply:
FYI in 1.12, we did some redesign so that it works with one scheduling cycle: https://github.com/kubernetes/kubernetes/pull/67556

Reply:
By the way, at the reserving phase, volume binding/provisioning needs to reserve PV/PVCs. Currently, it is done by [call assumeVolumes before assuming the pod](https://github.com/kubernetes/kubernetes/blob/b9631a881f327b58ff52654cbb448927226bf05a/pkg/scheduler/scheduler.go#L489-L501).